pilot/pkg/config/kube/gateway/conversion.go

```diff
@@ -1449,8 +1450,24 @@ func convertGateways(r ConfigContext) ([]config.Config, map[parentKey]map[k8s.Se
 			if len(addressesToReport) == 0 {
 				// There are no external addresses, so report the internal ones
 				// TODO: should we always report both?
-				addressesToReport = internal
 				addrType = k8s.HostnameAddressType
+				for _, hostport := range internal {
+					svchost, _, _ := net.SplitHostPort(hostport)
+					svc := r.Context.ps.ServiceIndex.HostnameAndNamespace[host.Name(svchost)][obj.Namespace]
```

Comment:
obj.Namespace is the namespace of the Gateway which only coincidentally aligns with the namespace of the hostname. This is another thing pointing to moving the logic into ResolveGatewayInstances